src/_pytest/assertion/util.py

```diff
@@ -189,7 +190,8 @@ def assertrepr_compare(
     explanation = None
     try:
         if op == "==":
-            explanation = _compare_eq_any(left, right, verbose)
+            writer = config.get_terminal_writer()
+            explanation = _compare_eq_any(left, right, writer, verbose)
```

Comment:
I don't like that we pass the full TerminalWriter to these functions, because it makes it seem like the functions actually write to the terminal, while they only need it for the pure `_highlight` function. But it's OK for now, not asking you to refactor :)

Reply:
If you prefer, we could instead pass a `highlight` function (`Callable[[str], str]`) which would be `TerminalWriter._highlight`

Reply:
I like that idea.

Reply:
Ok, I have pushed a fixup for this. I wasn't sure where to add the type definition, feel free to update/let me know if there's a better place

Reply:
I think that looks great, thanks!